### PR TITLE
Docker compose restart behaviour

### PR DESCRIPTION
**Pull request by zuzuleinen**
closes https://github.com/benchkram/bob/issues/98

This PR also fixes the issue with rebuild not happening after a restart

### CODE REVIEW

File: pkg/ctl/commander.go [hunk 1]
```diff
@@ -64,17 +67,17 @@ type Builder interface {
 // |___________|           |___________|           |___________|
 //
 func NewCommander(ctx context.Context, builder Builder, ctls ...Command) Commander {
-
-	c := &commander{
+	cmdr := &commander{
```
Comment (Equanox) [inline pkg/ctl/commander.go:70]: If you change it here `c -> cmdr` assure to also change it in the member function definitions `func (c *commander)  -> func (cmdr *commander) `

File: pkg/ctl/commander.go [hunk 2]
```diff
@@ -103,17 +106,14 @@ func NewCommander(ctx context.Context, builder Builder, ctls ...Command) Command
...
 						defer done()
 
-						err := c.Stop()
-						boblog.Log.Error(err, "Error on stopping comander")
+						err := cmdr.Restart()
+						boblog.Log.Error(err, "Error on restarting commander")
 
 						// Trigger a rebuild.
-						err = c.builder.Build(ctx)
+						err = cmdr.builder.Build(ctx)
```
Comment (Equanox) [inline pkg/ctl/commander.go:113]: Build() must happen before restarting.

**Comment by Equanox** [comment]
Check if the behavior is correct when a init is in progress, here https://github.com/benchkram/bob/blob/main/bobrun/wrap.go#L97

**Comment by Equanox** [comment]
A restart must be able to be cancable (by listening to the context as we do it with build).. E.g. A restart is immediately followed by a ctrl+c (send SIGINT).. this should stop the restart and allow shutdown to start as fast as possible.

File: pkg/ctl/commander.go (hunk 1 above)
Comment (zuzuleinen) [reply pkg/ctl/commander.go:70]: After taking a second look I think `c` is better. Will change back to that.

**Comment by zuzuleinen** [comment]
> A restart must be able to be cancable (by listening to the context as we do it with build).. E.g. A restart is immediately followed by a ctrl+c (send SIGINT).. this should stop the restart and allow shutdown to start as fast as possible.

It looks like it already is if I'm not mistaken. How I tested:

In /home/andrei/Projects/bob/pkg/execctl/cmd.go, I changed Restart() method with:

```go
// Restart first interrupts the command if it's already running, and then re-runs the command.
func (c *Cmd) Restart() error {
	err := c.stop()
	if err != nil {
		return err
	}

	err = c.Wait()
	if err != nil {
		return err
	}

	fmt.Println("sleeping")
	time.Sleep(20 * time.Second)
	fmt.Println("done")

	return c.Start()
}
```

On the bob/example/server-db I did a restart and CTRL+R.

Result:

The exec command from above will not start and then everything is shutting down.

**Comment by zuzuleinen** [comment]
Testing wait for init to finish before restarting

1. In example/server-db add init to the server run task:

```yaml
run:
  server:
    type: binary
    path: ./build/server
    dependson:
      - build
      - database
    init: "echo initstart\nsleep 10s\n echo donewithinit"
  database:
    type: compose
    path: docker-compose.yml

```

2. Add some logs in:

pkg/execctl/cmd.go
bobrun/compose.go 

where restart is happening.

2. Run `bob run server`

3. Hit CTRL-R to restart in the 10s interval while the init cmd is running
4. Should restart only after init is done.The ouput of init should be before the logs added in the cmd and compose

**Comment by Equanox** [comment]
Tested it as you described. Seems that restart is not waiting for init to be finished.. It also doesn't executes init again after subsequent restarts.

image

**Comment by Equanox** [comment]
I found another bug. `CTRL+R` seems not to trigger a rebuild. Do a change in `example/server-db/main.go` and do a restart.
`bob build` does a rebuild .. `CTRL+R` misses this.

**Comment by Equanox** [comment]
Closing for now, focus is currently on build

**Pull request closed without merging**